File: tools/xdhq/wrappers/ZND/XDHq.php
<?php
/*
	Copyright (C) 2018 by Claude SIMON (http://zeusw.org/epeios/contact.html).

	This file is part of XDHq.

	XDHq is free software: you can redistribute it and/or modify it
	under the terms of the GNU Affero General Public License as published
	by the Free Software Foundation, either version 3 of the License, or
	(at your option) any later version.

	XDHq is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	GNU Affero General Public License for more details.

	You should have received a copy of the GNU Affero General Public License
	along with XDHq.  If not, see <http://www.gnu.org/licenses/>.
*/
require 'XDHqSHRD.php';
require 'XDHqDEMO.php';
require 'XDHqXML.php';

if ( false && XDHq_SHRD::isDev() )
	require 'XDHqPROD.php';


class XDHq extends XDHq_SHRD{
	const MODE_PROD = 0;
	const MODE_DEMO = 1;
	const MODE_UNDEFINED = 2;
	private static $mode_ = self::MODE_UNDEFINED;
	static $dir;
	private static function getAssetPath_( $dir ) {
		if ( parent::isDev() )
			return "h:/hg/epeios/tools/xdhq/examples/common/" . $dir . "/";
		else {
			list($script)=get_included_files();
			return dirname($script) . "/";
		}
	}
	private static function getAssetFilename_( $path, $dir ) {
		return self::getAssetPath_( $dir ) . $path;
	}
	static function readAsset( $path, string $dir = "" ) {
		return file_get_contents( self::getAssetFilename_( $path, $dir ) );
	}
	static function launch( string $newSessionAction, string $headContent, $mode, string $dir ) {
		self::$mode_ = $mode;
		self::$dir = $dir;

		switch ( $mode ) {
		case self::MODE_PROD:
			XDHq_PROD::launch( $newSessionAction );
			break;
		case self::MODE_DEMO:
			XDHq_DEMO::launch( $newSessionAction, $headContent );
			break;
		default:
			throw new Exception( "Unknown mode !!!");
			break;
		}
	}
	static function getMode() {
		return self::$mode_;
	}
	static function isDEMO() {
		return self::getMode() == self::MODE_DEMO;
	}
}


class XDHqDOM extends Threaded {
	private $dom_;
	private function split_( array $keysAndValues, array &$keys, array &$values ) {
		foreach ($keysAndValues as $key => $value) {
			$keys[] = $key;
			$values[] = $value;
		}
	}
	private function unsplit_( array $keys, array $values ) {
		$count = count( $keys );
		$i = 0;
		$keysAndValues = [];

		while ( $i < $count ) {
			$keysAndValues[$keys[$i]] = $values[$i];
			$i++;
		}

		return $keysAndValues;
	}
	function __construct() {
		switch ( XDHq::getMode() ) {
		case XDHq::MODE_PROD:
			$this->dom_ = new XDHqDOM_PROD;
			break;
		case XDHq::MODE_DEMO:
			$this->dom_ = new XDHqDOM_DEMO;
			break;
		default:
			die( "Unknown mode !!!");
		}
	}
	private function call_( ...$args ) {
		return $this->dom_->call( ...$args );
	}
	function getAction( &$id ) {
		return $this->dom_->getAction( $id );
	}
	function execute( $script ) {
		return self::call_( "Execute_1", XDHq::RT_STRING, 1, $script, 0 );
	}
	function alert( string $message ) {
		self::call_( "Alert_1", XDHq::RT_NONE, 1, $message, 0 );
	}
	function confirm( string $message ) {
		return self::call_( "Confirm_1", XDHq::RT_STRING, 1, $message, 0 ) == "true";
	}
	private function setLayout_(string  $id, $xml, string $xslFilename ) {
		self::call_( "SetLayout_1", XDHq::RT_NONE, 3, $id, $xml, $xslFilename, 0 );
	}
	function setLayout(string  $id, string $html ) {
		self::setLayout_( $id, $html, "" );
	}
	function setLayoutXSL(string  $id, $xml, string $xsl ) {
		$xslURL = $xsl;

		if ( XDHq::isDEMO() )
			$xslURL = "data:text/xml;charset=utf-8," . rawurlencode( XDHq::readAsset( $xsl, XDHq::$dir ) );
			
		self::setLayout_( $id, is_string($xml) ? $xml : $xml->toString(), $xslURL );
	}
	function getContents( array $ids ) {
		return self::unsplit_($ids,self::call_( "GetContents_1", XDHq::RT_STRINGS, 0, 1, $ids ));
	}
	function getContent( string $id ) {
		return self::getContents( [$id] )[$id];
	}
	function setContents( array $idsAndContents ) {
		$ids = [];
		$contents = [];

		self::split_( $idsAndContents, $ids, $contents );

		self::call_( "SetContents_1", XDHq::RT_NONE, 0, 2, $ids, $contents );
	}
	function setContent( string $id, string $content ) {
		self::setContents( [ $id => $content ] );
	}
	function setTimeout( int $delay, string $action ) {
		self::call_( "SetTimeout_1", XDHq::RT_NONE, 2, strval( $delay ), $action, 0 );
	}
	function createElement( string $name, string $id = "" ) {
		return self::call_( "CreateElement_1", XDHq::RT_STRING, 2, $name, $id, 0 );
	}
	function insertChild( string $child, string $id ) {
		self::call_( "InsertChild_1", XDHq::RT_NONE, 2, $child, $id, 0 );
	}
	function dressWidgets( string $id ) {
		return self::call_( "DressWidgets_1", XDHq::RT_NONE, 1, $id, 0 );
	}
	private function handleClasses( $command, array $idsAndClasses ) {
		$ids = [];
		$classes = [];

		self::split_( $idsAndClasses, $ids, $classes );
		self::call_( $command, XDHq::RT_NONE, 0, 2, $ids, $classes );
	}
	function addClasses( array $idsAndClasses ) {
		self::handleClasses( "AddClasses_1", $idsAndClasses );
	}
	function removeClasses( array $idsAndClasses ) {
		self::handleClasses( "RemoveClasses_1", $idsAndClasses );
	}
	function toggleClasses( array $idsAndClasses ) {
		self::handleClasses( "ToggleClasses_1", $idsAndClasses );
	}
	function addClass( string $id, string $class  ) {
		self::addClasses( [ $id => $class ] );
	}
	function removeClass( string $id, string $class  ) {
		self::removeClasses( [ $id => $class ] );
	}
	function toggleClass( string $id, string $class  ) {
		self::toggleClasses( [ $id => $class ] );
	}
	function enableElements( array $ids ) {
		self::call_( "EnableElements_1", XDHq::RT_NONE, 0, 1, $ids );
	}
	function enableElement( string $id ) {
		self::enableElements( array( $id ) );
	}
	function disableElements( array $ids ) {
		self::call_( "DisableElements_1", XDHq::RT_NONE, 0, 1, $ids  );
	}
	function disableElement( string $id ) {
		self::disableElements( array( $id ) );
	}
	function setAttribute( string $id, string $name, string $value ) {
		return self::call_( "SetAttribute_1", XDHq::RT_NONE, 2, $id, $value, 0 );
	}
	function getAttribute( string $id, string $name ) {
		return self::call_( "GetAttribute_1", XDHq::RT_STRING, 2, $id, $name, 0 );
	}
	function removeAttribute( string $id, string $name ) {
		self::call_( "RemoveAttribute_1", XDHq::RT_NONE, 1, $id, 0 );
	}
	function setProperty( string $id, string $name, string $value ) {
		return self::call_( "SetProperty_1", XDHq::RT_NONE, 3, $id, $name, $value, 0 );
	}
	function getProperty( string $id, string $name ) {
		return self::call_( "GetProperty_1", XDHq::RT_STRING, 2, $id, $name, 0 );
	}
	function focus( string $id ) {
		self::call_( "Focus_1", XDHq::RT_NONE, 1, $id, 0 );
	}
}
?>

File: tools/xdhq/wrappers/PYH/XDHq.py
# coding: utf-8

"""
MIT License

Copyright (c) 2018 Claude SIMON (https://q37.info/s/rmnmqd49)

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

import XDHqDEMO, XDHqSHRD, XDHqXML

import os, sys
from collections import OrderedDict

if sys.version_info[0] == 2:
	import urllib
	def _encode(string):
		return urllib.quote(string)
elif sys.version_info[0] == 3:
	import urllib.parse
	def _encode(string):
		return urllib.parse.quote(string)
else:
	print("Unhandled python version!")
	os._exit(1)

_dir = ""

_VOID = XDHqSHRD.RT_VOID
_STRING = XDHqSHRD.RT_STRING
_STRINGS = XDHqSHRD.RT_STRINGS
XML = XDHqXML.XML

def _split(keysAndValues):
	keys = []
	values = []

	for key in keysAndValues:
		keys.append(str(key))
		values.append(str(keysAndValues[key]))

	return [keys,values]

def _unsplit(keys,values):
	i = 0
	# With 'OrderedDict', the order of the items is keeped under Python 2.
	# This facilitates the retrieving of values by using 'values()' method.
	# Dict are ordered by default under Python 3.
	keysAndValues = OrderedDict()
	length = len(keys)

	while i < length:
		keysAndValues[keys[i]] = values[i]
		i += 1

	return keysAndValues

def _getAssetPath(dir):
	if XDHqSHRD.isDev():
		return os.path.join(os.environ["Q37_EPEIOS"],"tools/xdhq/examples/common/", dir )
	else:
		return os.path.abspath(os.path.dirname(sys.argv[0]))

def _getAssetFilename(path, dir):
	return os.path.join(_getAssetPath(dir), path )

def readAsset(path, dir=""):
	return open(_getAssetFilename(path, dir)).read()

def _readXSLAsset(path, dir):
	if (path.lstrip()[0]=='<'):
		return path.lstrip()
	else:
		return readAsset(path, dir)

class DOM:
	def __init__(self,instance):
		self._dom = XDHqDEMO.DOM_DEMO(instance)

	def getAction(self):
		return self._dom.getAction()

	def execute(self,script):
		return self._dom.call("Execute_1" ,_STRING, script)

	def flush(self):	# Returns when all the pending command were executed.
		self.execute("''")

	def alert(self,message):
		self._dom.call( "Alert_1", _STRING, message )
		# For the return value being 'STRING' instead of 'VOID',
		# see the 'alert' primitive in 'XDHqXDH'.

	def confirm(self,message):
		return self._dom.call( "Confirm_1", _STRING, message ) == "true"

	def _handleLayout(self, variant, id, xml, xsl):
		#	If 'xslFilename' is empty, 'xml' contents HTML.
		# 	If 'xml' is HTML and uses the compressed form, if it has a root tag, only the children will be used.
		self._dom.call("HandleLayout_1", _VOID, variant, id, xml if isinstance(xml,str) else xml.toString(), xsl)

	def prependLayout(self,id,html):
		self._handleLayout("Prepend",id,html,"")

	def setLayout(self,id,html):
		self._handleLayout("Set",id,html,"")

	def appendLayout(self,id,html):
		self._handleLayout("Append",id,html,"")

	def _handleLayoutXSL(self, variant, id, xml, xsl):
		global _dir
		xslURL = xsl

		if True:	# Testing if 'PROD' or 'DEMO' mode when available.
			xslURL = "data:text/xml;charset=utf-8," + _encode( _readXSLAsset( xsl, _dir ) )

		self._handleLayout(variant, id, xml, xslURL )

	def prependLayoutXSL(self, id, xml, xsl):
		self._handleLayoutXSL("Prepend",id,xml,xsl)

	def setLayoutXSL(self, id, xml, xsl):
		self._handleLayoutXSL("Set",id,xml,xsl)

	def appendLayoutXSL(self, id, xml, xsl):
		self._handleLayoutXSL("Append",id,xml,xsl)

	def getContents(self, ids):
		return _unsplit(ids,self._dom.call("GetContents_1",_STRINGS, ids))

	def getContent( self, id):
		return self.getContents([id])[id]

	def setContents(self,idsAndContents):
		[ids,contents] = _split(idsAndContents)

		self._dom.call("SetContents_1", _VOID, ids, contents)

	def setContent(self, id, content):
		self.setContents({id: content})

	def setTimeout(self,delay,action ):
		self._dom.call( "SetTimeout_1", _VOID, str( delay ), action )

	"""
	# Following 4 methods will either be removed or redesigned.

	# Will become a variation of 'createElementNS(…)',
	# with a optional list of attributes.
	def createElement(self, name, id = "" ):
		return self._dom.call( "CreateElement_1", _STRING, 2, name, id, 0 )

	# Will become 'prependChild(…)', with variations.
	def insertChild(self,child,id):
		self._dom.call( "InsertChild_1", _VOID, 2, child, id, 0 )

	# NOTA: The 'CSSRule' related methods will be probably removed.
	# Enabling/disabling styles are easier to use.
	def insertCSSRule(self,rule,index,id=""):
		self._dom.call("InsertCSSRule_1", _VOID, 3, id, rule, str(index), 0)

	def appendCSSRule(self,rule,id=""):
		return int(self._dom.call("AppendCSSRule_1", _STRING, 2, id, rule, 0))

	def removeCSSRule(self,index,id=""):
		self._dom.call("RemoveCSSRule_1", _VOID, 2, id, str(index), 0)
	"""

	def dressWidgets(self,id):
		return self._dom.call( "DressWidgets_1", _VOID, id )

	def _handleClasses(self, variant, idsAndClasses):
		[ids, classes] = _split(idsAndClasses)

		self._dom.call("HandleClasses_1", _VOID, variant, ids, classes)

	def addClasses(self, idsAndClasses):
		self._handleClasses("Add", idsAndClasses)

	def removeClasses(self, idsAndClasses):
		self._handleClasses("Remove", idsAndClasses)

	def toggleClasses(self, idsAndClasses):
		self._handleClasses("Toggle", idsAndClasses)

	def addClass(self, id, clas ):
		self.addClasses({id: clas})

	def removeClass(self, id, clas ):
		self.removeClasses({id: clas})

	def toggleClass(self, id, clas ):
		self.toggleClasses({id: clas})

	def enableElements(self,ids):
		self._dom.call("EnableElements_1", _VOID, ids )

	def enableElement(self, id):
		self.enableElements([id] )

	def disableElements(self, ids):
		self._dom.call("DisableElements_1", _VOID, ids )

	def disableElement(self, id):
		self.disableElements([id])

	def setAttribute(self, id, name, value ):
		self._dom.call("SetAttribute_1", _VOID, id, name, str(value) )

	def getAttribute(self, id, name):
		return self._dom.call("GetAttribute_1", _STRING, id, name )

	def removeAttribute(self, id, name ):
		self._dom.call("RemoveAttribute_1", _VOID, id, name )

	def setProperty(self, id, name, value ):
		self._dom.call("SetProperty_1", _VOID, id, name, value )

	def getProperty(self, id, name ):
		return self._dom.call("GetProperty_1", _STRING, id, name )

	def focus(self, id):
		self._dom.call("Focus_1", _VOID, id)

def launch(callback, userCallback, callbacks, headContent, dir):
	global _dir
	_dir = dir
	XDHqDEMO.launch(callback, userCallback,callbacks,headContent)
